### DISCUSSION TITLE
Recommended best practice for runtime linking against libraries installed via Spack?

### DISCUSSION THREAD

**Discussion by carns** [spack/spack#32192] [General]
What is the recommended best practice for setting correct paths for runtime dynamic linking against libraries installed via Spack?  I'm referring specifically to the case where the application using those libraries is not itself using Spack.

We have this use case in the Mochi project.  We provide a number of small composable libraries (each is a separate Spack package, maintained in an external package repo) that end-users/customers (developers of data services) might mix and match for their own needs.  We don't control what build system they use; we are just providing libraries and headers.

Prior to https://github.com/spack/spack/pull/28354,  it was trivial for codes built against our libraries to find them at runtime, because Spack would put all of the library paths in the LD_LIBRARY_PATH variable.  I understand why the change in that PR was made, though.  The purpose of this discussion thread isn't to debate the underlying policy but rather to learn best practice for how to operate with it.

Here are the options that I am aware of:

- Have our end-users put a `modules:` section in their environment files to cause LD_LIBRARY_PATH to be set for all loaded packages as it was before.
- If you are using autotools/libtool, add explicit rpath/runpath references to external libraries using the `-R` option to libtool, as in this proof of concept: https://github.com/mochi-hpc/mochi-margo/pull/196/files.  This ties the application to a specific concretized library install.
- Rely on .la files in the spack installation directory to help inform libtool so that it knows when rpath/runpath is needed.  I don't see how to do this in Spack, though; Spack explicitly scrubs these in a post-install step.
- Place the libraries in a standard path that the runtime linker already searches.  This obviously isn't compatible with the Spack installation philosophy, though.

I think CMake behavior is different than Autotools behavior in terms of how it reacts to dependent libraries in non-standard paths, but I'm afraid that I don't know enough about CMake to understand the difference :)  My understanding is that autotools/libtool is very conservative; if you link (at compile time) against a library in a non-standard location, it assumes that either you are going to put it in a standard location later or provide a .la file that gives it better hints about what to do.  I think CMake is likely more aggressive about this, or at least makes it easier to force rpath/runpath settings.  We can't control what build systems our customers use, though.

My current best guess is to recommend that our uses do either the 1st or 2nd options above, but I feel like I'm missing something about how this is supposed to work.  Any suggestions?

**Comment by salotz** [comment]
I do something similar for applications that will not ever get their own Spack packages. I typically create an environment with a view and then you just need to set the environment configuration to that prefix. So I would say 1 or 4 above in your list depending on how the user would like to do things. For instance they might have a single environment view they use across many different projects and so they could put it in `~/.spack/envs/dev` and then use `ldconfig` to add this to the list of libraries or they might just set the `LD_LIBRARY_PATH`.

In any case, I'm not sure there is a way for Spack to enforce that kind of behavior nor would it be desirable as many will have their own desired way of working or deploying things.

I think if you wanted a more "Spack" solution then you would have to create a package recipe for the applications and have them part of the solve.

**Comment by haampie** [comment]
Another solution is here: https://github.com/spack/spack/pull/31948 it patches the soname of shared libraries post-install to their install paths, so that even if you're linking to spack-built libraries by hand, you'll register the needed library by path instead of by name -- no dynamic loader search required anymore.

**Reply by salotz** [reply]
So from a consumer's perspective they would have to provide the compiler flags when compiling their projects? How would this look in practice using a build system?
